Mathlib/Data/Fintype/EquivFin.lean

```diff
@@ -0,0 +1,688 @@
+-/
+noncomputable def equivFin (α) [Fintype α] : α ≃ Fin (card α) :=
+  letI := Classical.decEq α
+  (truncEquivFin α).out
+
+/-- There is (computably) a bijection between `Fin (card α)` and `α`.
+
+Since it is not unique and depends on which permutation
+of the universe list is used, the bijection is wrapped in `Trunc` to
+preserve computability.
+
+See `Fintype.truncEquivFin` for a version that gives an equivalence
+given `[DecidableEq α]`.
+-/
+def truncFinBijection (α) [Fintype α] : Trunc { f : Fin (card α) → α // Bijective f } := by
+  unfold card Finset.card
+  refine
+    Quot.recOnSubsingleton
+      (motive := fun s : Multiset α =>
+        (∀ x : α, x ∈ s) → s.Nodup → Trunc {f : Fin (Multiset.card s) → α // Bijective f})
+      univ.val
+      (fun l (h : ∀ x : α, x ∈ l) (nd : l.Nodup) => Trunc.mk (nd.getBijectionOfForallMemList _ h))
+      mem_univ_val univ.2
+
+end Fintype
+
+namespace Fintype
+
+section
+
+variable [Fintype α] [Fintype β]
+
+/-- If the cardinality of `α` is `n`, there is computably a bijection between `α` and `Fin n`.
+
+See `Fintype.equivFinOfCardEq` for the noncomputable definition,
+and `Fintype.truncEquivFin` and `Fintype.equivFin` for the bijection `α ≃ Fin (card α)`.
+-/
+def truncEquivFinOfCardEq [DecidableEq α] {n : ℕ} (h : Fintype.card α = n) : Trunc (α ≃ Fin n) :=
+  (truncEquivFin α).map fun e => e.trans (finCongr h)
+
+/-- If the cardinality of `α` is `n`, there is noncomputably a bijection between `α` and `Fin n`.
+
+See `Fintype.truncEquivFinOfCardEq` for the computable definition,
+and `Fintype.truncEquivFin` and `Fintype.equivFin` for the bijection `α ≃ Fin (card α)`.
+-/
+noncomputable def equivFinOfCardEq {n : ℕ} (h : Fintype.card α = n) : α ≃ Fin n :=
+  letI := Classical.decEq α
+  (truncEquivFinOfCardEq h).out
+
+/-- Two `Fintype`s with the same cardinality are (computably) in bijection.
+
+See `Fintype.equivOfCardEq` for the noncomputable version,
+and `Fintype.truncEquivFinOfCardEq` and `Fintype.equivFinOfCardEq` for
+the specialization to `Fin`.
+-/
+def truncEquivOfCardEq [DecidableEq α] [DecidableEq β] (h : card α = card β) : Trunc (α ≃ β) :=
+  (truncEquivFinOfCardEq h).bind fun e => (truncEquivFin β).map fun e' => e.trans e'.symm
+
+/-- Two `Fintype`s with the same cardinality are (noncomputably) in bijection.
+
+See `Fintype.truncEquivOfCardEq` for the computable version,
+and `Fintype.truncEquivFinOfCardEq` and `Fintype.equivFinOfCardEq` for
+the specialization to `Fin`.
+-/
+noncomputable def equivOfCardEq (h : card α = card β) : α ≃ β := by
+  letI := Classical.decEq α
+  letI := Classical.decEq β
+  exact (truncEquivOfCardEq h).out
+
+end
+
+theorem card_eq {α β} [_F : Fintype α] [_G : Fintype β] : card α = card β ↔ Nonempty (α ≃ β) :=
+  ⟨fun h =>
+    haveI := Classical.propDecidable
+    (truncEquivOfCardEq h).nonempty,
+    fun ⟨f⟩ => card_congr f⟩
+
+end Fintype
+
+@[simp]
+theorem Fintype.card_fin (n : ℕ) : Fintype.card (Fin n) = n :=
+  List.length_finRange n
```

Comment:
Shouldn't that be closer to the `Fintype (Fin n)` instance?

Reply:
We can indeed upstream it to `Fintype/Card.lean` itself, since that imports `Fintype/Basic.lean` which holds the `Fintype (Fin n)` instance.